File: sgp/sgp/custom/py/job_card.py
import frappe
from frappe.desk.form import assign_to

def create_timesheet(doc,action):
    time_logs = doc.time_logs
    employees = list(set([i.employee for i in time_logs]))
    for i in employees:
        timesheet = frappe.new_doc("Timesheet")
        if(not frappe.db.exists("Activity Type", "Production")):
            activity = frappe.new_doc("Activity Type")
            activity.activity_type = "Production"
            activity.save()
        employees_time_log = [j for j in time_logs if(j.employee == i)] 
        timelogs_timesheet = []
        for k in employees_time_log:
            timelogs_timesheet.append({
                'activity_type': "Production",
                'from_time' : k.from_time,
                'to_time' : k.to_time,
                'hours' :  k.time_in_mins/60,
                'total_production_pavers' : k.completed_qty
            })
        timesheet.update({
            'company': doc.company,
            'workstation' : doc.workstation,
            'time_logs' : timelogs_timesheet,
            'employee' : i
        })
        timesheet.insert(ignore_permissions = True)
        timesheet.submit()
        user_id = frappe.get_value("Employee", i, 'user_id')
        if(user_id):
            assign_to.add(
                    dict(
                        assign_to=[user_id],
                        doctype="Timesheet",
                        name=timesheet.name,
                        notify=True,
                    )
                )
        else:
            frappe.msgprint(f"User id not set for Employee: {i}")
    
@frappe.whitelist()
def get_workorder_doc(work_order, opr, workstation, qty=0):
    wo=frappe.get_doc("Work Order", work_order)
    over_prdn_prcnt = frappe.db.get_singles_value("Manufacturing Settings", 'overproduction_percentage_for_work_order')
    wo.update({
        'over_prdn_prcnt' : over_prdn_prcnt
    })
    frappe.db.set_value("Work Order Operation",{'operation':opr,'workstation':workstation,'parent':work_order},'completed_qty', qty)
    frappe.db.commit()
    return wo

File: sgp/sgp/custom/js/job_card.js
frappe.ui.form.on("Job Card",{
    refresh: function(frm){
        frm.add_custom_button("Finish",()=>{
            if(frm.doc.work_order){
            frappe.call({
                method:"sgp.sgp.custom.py.job_card.get_workorder_doc",
                args:{
                    work_order:frm.doc.work_order,
                    opr:frm.doc.operation,
                    workstation:frm.doc.workstation,
                    qty : frm.doc.total_completed_qty
                    },
                callback(r){
                    make_se(r.message, "Manufacture")
                }
            })
        }
        }).addClass("btn-primary")
    },
    onload: function(frm){
        if(frm.doc.doc_onload == 0){
        frm.set_value('time_logs',[])
        frm.set_value('doc_onload',1)
        frm.refresh()
        frm.save()
        }
    }
})
function make_se (frm, purpose) {
    show_prompt_for_qty_input(frm, purpose)
        .then(data => {
            if(data.qty<=0){return}
            return frappe.xcall('erpnext.manufacturing.doctype.work_order.work_order.make_stock_entry', {
                'work_order_id': frm.name,
                'purpose': purpose,
                'qty': data.qty
            });
        }).then(stock_entry => {
            frappe.model.sync(stock_entry);
            frappe.set_route('Form', stock_entry.doctype, stock_entry.name);
        });
        

}
function show_prompt_for_qty_input(frm, purpose) {
    let max = get_max_transferable_qty(frm, purpose);
    return new Promise((resolve, reject) => {
        frappe.prompt({
            fieldtype: 'Float',
            label: __('Qty for {0}', [purpose]),
            fieldname: 'qty',
            description: __('Max: {0}', [max]),
            default: max
        }, data => {
            max += (frm.qty * (frm.over_prdn_prcnt || 0.0)) / 100;

            if (data.qty > max) {
                frappe.msgprint(__('Quantity must not be more than {0}', [max]));
                reject();
            }
            if(data.qty > cur_frm.doc.total_completed_qty)
            data.purpose = purpose;
            resolve(data);
        }, __('Select Quantity'), __('Create'));
    });
}
function get_max_transferable_qty (frm, purpose){
    let max = 0;
    if (frm.skip_transfer) {
        max = flt(frm.qty) - flt(frm.produced_qty);
    } else {
        if (purpose === 'Manufacture') {
            max = flt(frm.material_transferred_for_manufacturing) - flt(frm.produced_qty);
        } else {
            max = flt(frm.qty) - flt(frm.material_transferred_for_manufacturing);
        }
    }
    return flt(max);
}